packages/remix-react/transition.ts

```diff
@@ -1053,14 +1053,18 @@ export function createTransitionManager(init: TransitionManagerInit) {
     let controller = new AbortController();
     pendingNavigationController = controller;
 
+    // Create a local copy we can mutate for proper determination of the acton
+    // to run on layout/index routes.  We do not want to mutate the eventual
+    // matches used for revalidation
+    let actionMatches = matches;
     if (
-      !isIndexRequestAction(submission.action) &&
-      matches[matches.length - 1].route.id.endsWith("/index")
+      !isIndexRequestUrl(createUrl(submission.action)) &&
+      actionMatches[matches.length - 1].route.index
     ) {
-      matches = matches.slice(0, -1);
```

Comment:
This as the bug - by mutating in place prior to getting the `leafMatch` we were causing the trimmed `matches` to persist through revalidation

Reply:
I also updated this to use the new `isIndexRequestUrl` which was doing the same thing as `isIndexRequestAction`